cleverhans/attacks.py

```diff
@@ -1,4 +1,342 @@
+
+        # Generate this attack's graph if it hasn't been done previously
+        if not hasattr(self, "_x"):
+            input_shape = list(X.shape)
+            input_shape[0] = None
+            self._x = tf.placeholder(tf.float32, shape=input_shape)
+            self._x_adv = self.generate(self._x)
+
+        # Run symbolic graph without or with true labels
+        if params['Y'] is None:
+            feed_dict = {self._x: X}
+        else:
+            feed_dict = {self._x: X, self.y: params['Y']}
+        return self.sess.run(self._x_adv, feed_dict=feed_dict)
+
+
+class BasicIterativeMethod(Attack):
+    """
+    The Basic Iterative Method (Kurakin et al. 2016). The original paper used
+    hard labels for this attack; no label smoothing.
+    Paper link: https://arxiv.org/pdf/1607.02533.pdf
+    """
+    def __init__(self, model, back='tf', sess=None, params={'eps': 0.05,
+                                                            'nb_iter': 10,
+                                                            'y': None,
+                                                            'ord': 'np.inf',
+                                                            'clip_min': None,
+                                                            'clip_max': None}):
+        """
+        Create a BasicIterativeMethod instance.
+
+        Attack-specific parameters:
+        :param eps: (required float) step size for each attack iteration
+        :param nb_iter: (required int) Number of attack iterations.
+        :param y: (required) A placeholder for the model labels.
+        :param ord: (optional) Order of the norm (mimics Numpy).
+                    Possible values: np.inf, 1 or 2.
+        :param clip_min: (optional float) Minimum input component value
+        :param clip_max: (optional float) Maximum input component value
+        """
+        super(BasicIterativeMethod, self).__init__(model, back, sess, params)
+
+        # Check that all required attack specific parameters are defined
+        req = ('eps', 'nb_iter', 'y')
+        if not all(k in params for k in req):
+            raise Exception("Attack requires label placeholder.")
+
+        # Save attack-specific parameters
+        self.eps = params['eps']
+        self.nb_iter = params['nb_iter']
+        self.y = params['y']
+        self.ord = params['ord'] if 'ord' in params else np.inf
+        self.clip_min = params['clip_min'] if 'clip_min' in params else None
+        self.clip_max = params['clip_max'] if 'clip_max' in params else None
+
+        # Check if order of the norm is acceptable given current implementation
+        if self.ord not in [np.inf, 1, 2]:
+            raise Exception("Norm order must be either np.inf, 1, or 2.")
+        if back == 'th':
+            error_string = "BasicIterativeMethod is not implemented in Theano"
+            raise NotImplementedError(error_string)
+
+    def generate(self, x, params={}):
+        import tensorflow as tf
```

Comment:
there is one in init so I did not put one here. Am happy to put it back for consistency if you think it's needed